Algorithm.md

```diff
@@ -0,0 +1,30 @@
+1. Fetch available versions for each dependency
+2. If a project is already in the graph, intersect the new requirement with the one already present
+3. Select the highest [precedence] version which satisfies the intersected requirement (or only requirement, if this is the first insertion)
+  * This always has to be O(n) over the list of versions, because custom requirements mean that a previously failing version might become a candidate again
+4. Collect all child dependencies from that graph depth, repeat from step 1
+
+### In event of a conflict
+
+Given a project already in the graph with requirement X, and about to be re-added to the graph with requirement Y, where X and Y conflict:
+
+1. Find the _dependent_ project which changed the requirement into X (either from nil, or from a previously looser requirement)
+2. Select the next best version of that dependent project and attempt to finish resolving
+3. In the event that no other versions satisfy the requirements applied to the dependent, we have two choices:
+  - GOTO step 1 for the dependent project, or
+  - Find a further back time where the originally conflicting dependency (the one with requirement X) was added (e.g., with requirement W), and try replacing the dependent which added it at _that_ level instead
```

Comment:
I'd suggest reading chapter 4 of [_Constraint Processing_](https://www.amazon.com/gp/product/1558608907/ref=as_li_tl?ie=UTF8&camp=1789&creative=9325&creativeASIN=1558608907&linkCode=as2&tag=mattdiephouse-20&linkId=1c958d2684749921ee5fba75c45c49a7).

This is equivalent to ordering the graph to minimize the induced width. That's tricky here, because the dependencies aren't set in stone, but I think it'd still be helpful to think about. That book, and chapter 4 specifically, has built up a lot of intuition for me about search spaces. But YMMV.